File: src/strategyClases.js
// ========================================
// PATRÓN STRATEGY - IMPLEMENTACIÓN CON CLASES EN JAVASCRIPT
// ========================================
// Este código muestra la implementación del patrón Strategy en JavaScript:
// - Cada estrategia de ordenamiento está encapsulada en su propia clase
// - El contexto puede cambiar dinámicamente entre estrategias
// - Cumple con el principio Open/Closed (abierto para extensión, cerrado para modificación)
// - Usa herencia de clases ES6 para implementar el patrón
// ========================================

// ========================================
// CLASE BASE DE ESTRATEGIA (STRATEGY BASE CLASS)
// ========================================
// Clase abstracta que define la interfaz común para todas las estrategias.
// En JavaScript no tenemos interfaces nativas, pero podemos simular
// el comportamiento con una clase base que lanza error si no se implementa.
// ========================================
class SortStrategy {
  // Método que debe ser implementado por las clases hijas
  sort(data) {
    throw new Error("Método sort() debe ser implementado"); // Si no se implementa, se lanza un error
  }
}

// ========================================
// ESTRATEGIAS CONCRETAS (CONCRETE STRATEGIES)
// ========================================
// Cada clase extiende SortStrategy e implementa su propia lógica
// de ordenamiento. Todas son intercambiables en tiempo de ejecución.
// ========================================

// Estrategia: Ordenamiento ascendente (menor a mayor)
class AscSortStrategy extends SortStrategy {
  sort(data) {
    // Usar spread operator para crear una copia del array original
    // y no mutar los datos originales (inmutabilidad)
    return [...data].sort((a, b) => a - b);
  }
}

// Estrategia: Ordenamiento descendente (mayor a menor)
class DescSortStrategy extends SortStrategy {
  sort(data) {
    return [...data].sort((a, b) => b - a); // copia para no mutar original
  }
}

// Estrategia: Sin ordenamiento (mantener orden original)
class NoSortStrategy extends SortStrategy {
  sort(data) {
    return [...data]; // copia para no mutar original
  }
}

// ========================================
// CONTEXTO (CONTEXT)
// ========================================
// El contexto mantiene una referencia a una estrategia
// y delega la ejecución del algoritmo en ella.
// Puede cambiar dinámicamente entre diferentes estrategias.
// ========================================
class SortContext {
  // Constructor: inicializa el contexto con una estrategia
  constructor(strategy) {
    this.strategy = strategy;
  }

  // Método para cambiar la estrategia en tiempo de ejecución
  setStrategy(strategy) {
    this.strategy = strategy;
  }

  // Método que ejecuta la estrategia actual
  // Delega la responsabilidad del ordenamiento a la estrategia
  sort(data) {
    return this.strategy.sort(data);
  }
}

// ========================================
// EJEMPLO DE USO (COMENTADO)
// ========================================
// El ejemplo de uso se comenta ya que lo estamos usando en el index.html
// para evitar duplicación de código. Aquí se muestra cómo se usaría:
// ========================================
/*
// Ejemplo de uso del patrón Strategy:

const data = [5, 3, 8, 1];

const context = new SortContext(new AscSortStrategy());
console.log('Ascendente:', context.sort(data));

// Cambiar a estrategia descendente (sin crear nuevo contexto)
context.setStrategy(new DescSortStrategy());
console.log('Descendente:', context.sort(data));

// Cambiar a estrategia sin ordenamiento
context.setStrategy(new NoSortStrategy());
console.log('Sin ordenar:', context.sort(data));
*/

File: src/strategyClases.cs
// ========================================
// PATRÓN STRATEGY - IMPLEMENTACIÓN CON CLASES
// ========================================
// Este código muestra la solución al problema anterior usando el patrón Strategy:
// - Cada estrategia de ordenamiento está encapsulada en su propia clase
// - El contexto puede cambiar dinámicamente entre estrategias
// - Cumple con el principio Open/Closed (abierto para extensión, cerrado para modificación)
// ========================================

using System;
using System.Collections.Generic;
using System.Linq;

// ========================================
// INTERFAZ DE ESTRATEGIA (STRATEGY INTERFACE)
// ========================================
// Define el contrato que todas las estrategias deben implementar.
// Esto permite que el contexto trabaje con cualquier estrategia
// que implemente esta interfaz (polimorfismo).
// ========================================
interface ISortStrategy
{
    List<int> Sort(List<int> data);
}

// ========================================
// ESTRATEGIAS CONCRETAS (CONCRETE STRATEGIES)
// ========================================
// Cada clase implementa una estrategia específica de ordenamiento.
// Todas implementan la interfaz ISortStrategy, por lo que son
// intercambiables en tiempo de ejecución.
// ========================================

// Estrategia: Ordenamiento ascendente (menor a mayor)
class AscSortStrategy : ISortStrategy
{
    public List<int> Sort(List<int> data)
    {
        return data.OrderBy(x => x).ToList();
    }
}

// Estrategia: Ordenamiento descendente (mayor a menor)
class DescSortStrategy : ISortStrategy
{
    public List<int> Sort(List<int> data)
    {
        return data.OrderByDescending(x => x).ToList();
    }
}

// Estrategia: Sin ordenamiento (mantener orden original)
class NoSortStrategy : ISortStrategy
{
    public List<int> Sort(List<int> data)
    {
        return new List<int>(data); // copia, sin ordenar
    }
}

// ========================================
// CONTEXTO (CONTEXT)
// ========================================
// El contexto mantiene una referencia a una estrategia
// y delega la ejecución del algoritmo en ella.
// Puede cambiar dinámicamente entre diferentes estrategias.
// ========================================
class SortContext
{
    // Referencia privada a la estrategia actual
    private ISortStrategy _strategy;

    // Constructor: inicializa el contexto con una estrategia
    public SortContext(ISortStrategy strategy)
    {
        _strategy = strategy;
    }

    // Método para cambiar la estrategia en tiempo de ejecución
    public void SetStrategy(ISortStrategy strategy)
    {
        _strategy = strategy;
    }

    // Método que ejecuta la estrategia actual
    // Delega la responsabilidad del ordenamiento a la estrategia
    public List<int> ExecuteSort(List<int> data)
    {
        return _strategy.Sort(data);
    }
}

// ========================================
// PROGRAMA PRINCIPAL - DEMOSTRACIÓN DEL USO
// ========================================
// Aquí se muestra cómo usar el patrón Strategy:
// 1. Crear un contexto con una estrategia inicial
// 2. Cambiar dinámicamente entre estrategias
// 3. Ejecutar cada estrategia con los mismos datos
// ========================================
class Program
{
    static void Main()
    {
        var data = new List<int> { 5, 3, 8, 1 };

        // ========================================
        // DEMOSTRACIÓN DEL PATRÓN STRATEGY
        // ========================================
        
        // Crear contexto con estrategia ascendente
        var context = new SortContext(new AscSortStrategy());
        Console.WriteLine("Ascendente: " + string.Join(", ", context.ExecuteSort(data)));

        // Cambiar a estrategia descendente (sin crear nuevo contexto)
        context.SetStrategy(new DescSortStrategy());
        Console.WriteLine("Descendente: " + string.Join(", ", context.ExecuteSort(data)));

        // Cambiar a estrategia sin ordenamiento
        context.SetStrategy(new NoSortStrategy());
        Console.WriteLine("Sin ordenar: " + string.Join(", ", context.ExecuteSort(data)));
    }
}
